File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int answer = 0;
    public boolean[] isVisited;
    
    public int solution(int k, int[][] dungeons) {
        
        isVisited = new boolean[dungeons.length];
        dfs(0, k, dungeons);
        
        return answer;
    }
    
    public void dfs(int cnt, int k, int[][] dungeons) {
        for (int i=0; i<dungeons.length; i++) {
            if (!isVisited[i] && k >= dungeons[i][0]) {
                isVisited[i] = true;
                dfs(cnt+1, k - dungeons[i][1], dungeons);
                isVisited[i] = false;
            }
        }
        
        answer = Math.max(answer, cnt);
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    answer = 0
    n = len(dungeons)
    for i in range(n):
        visited = [0] * n
        answer = max(answer, dfs(i, visited, k, dungeons, 0))
    
    return answer


def dfs(idx, visited, k, dungeons, answer):
    
    if not (k >= dungeons[idx][0] and k-dungeons[idx][1] >= 0):
        return 0
    
    
    visited[idx] = 1
    answer += 1
    k -= dungeons[idx][1]
    max_answer = answer
    
    
    for i in range(len(visited)):
        if visited[i] == 0:
            max_answer = max(max_answer, dfs(i, visited[:], k, dungeons, answer))
        
    return max_answer